misk-web/web/@misk/components/src/ducks/networkDucks.ts

```diff
@@ -0,0 +1,340 @@
+    yield put(dispatchNetwork.failure({ [tag]: { error: { ...e } } }))
+  }
+}
+
+function* handlePatch(
+  action: IAction<
+    NETWORK,
+    {
+      tag: string
+      url: string
+      updateData: any
+      requestConfig: AxiosRequestConfig
+    }
+  >
+) {
+  try {
+    const { tag, url, updateData, requestConfig } = action.payload
+    const { data } = yield call(axios.patch, url, { updateData }, requestConfig)
+    yield put(dispatchNetwork.success({ [tag]: data }))
+  } catch (e) {
+    const { tag } = action.payload
+    yield put(dispatchNetwork.failure({ [tag]: { error: { ...e } } }))
+  }
+}
+
+function* handlePost(
+  action: IAction<
+    NETWORK,
+    {
+      tag: string
+      url: string
+      saveData: any
+      requestConfig: AxiosRequestConfig
+    }
+  >
+) {
+  try {
+    const { tag, url, saveData, requestConfig } = action.payload
+    const { data } = yield call(axios.post, url, { saveData }, requestConfig)
+    yield put(dispatchNetwork.success({ [tag]: data }))
+  } catch (e) {
+    const { tag } = action.payload
+    yield put(dispatchNetwork.failure({ [tag]: { error: { ...e } } }))
+  }
+}
+
+function* handlePut(
+  action: IAction<
+    NETWORK,
+    {
+      tag: string
+      url: string
+      updateData: any
+      requestConfig: AxiosRequestConfig
+    }
+  >
+) {
+  try {
+    const { tag, url, updateData, requestConfig } = action.payload
+    const { data } = yield call(axios.put, url, { updateData }, requestConfig)
+    yield put(dispatchNetwork.success({ [tag]: data }))
+  } catch (e) {
+    const { tag } = action.payload
+    yield put(dispatchNetwork.failure({ [tag]: { error: { ...e } } }))
+  }
+}
+
+export function* watchNetworkSagas() {
+  yield all([
+    takeEvery(NETWORK.DELETE, handleDelete),
+    takeEvery(NETWORK.GET, handleGet),
+    takeEvery(NETWORK.PATCH, handlePatch),
+    takeEvery(NETWORK.POST, handlePost),
+    takeEvery(NETWORK.PUT, handlePut)
+  ])
+}
+
+/**
+ * Initial State
+ * Reducer merges all changes from dispatched action objects on to this initial state
+ */
+const initialState = fromJS({
+  ...defaultState.toJS()
+})
+
+/**
+ * Duck Reducer
+ * Merges dispatched action objects on to the existing (or initial) state to generate new state
+ */
+export function NetworkReducer(
+  state = initialState,
+  action: IAction<NETWORK, {}>
+) {
+  switch (action.type) {
+    case NETWORK.DELETE:
+    case NETWORK.FAILURE:
+    case NETWORK.GET:
+    case NETWORK.PATCH:
+    case NETWORK.POST:
+    case NETWORK.PUT:
+    case NETWORK.SUCCESS:
+      return state.mergeDeep(action.payload)
+    default:
+      return state
+  }
+}
+
+/**
+ * State Interface
+ * Provides a complete Typescript interface for the object on state that this duck manages
+ * Consumed by the root reducer in ./ducks index to update global state
+ * Duck state is attached at the root level of global state
+ */
+export interface INetworkState extends IDefaultState {
+  network: {
+    [tag: string]: {
```

Comment:
Wouldn't this be better represented as `[tag: string]: { data: any | null } | { error: any | null }`? Or do you anticipate you could have a network result with both?

Reply:
I don't anticipate a case where there is both a network result and an error. Though the approach you described would change the logic since there no longer is the assumption that the data and error keys exist.

Thus the current use of the state...
```Javascript
if (this.props.network.tag.data) {}
```

would have to change to...
```Javascript
if (this.props.network.tag.containsKey("data") && this.props.network.tag.data) {}
```

Given that the Redux pattern we use already does have some default keys that can be assumed to be present [(data, error, loading, success)](https://github.com/square/misk-web/blob/master/misk-web/web/%40misk/common/src/index.ts#L60), I'm leaning towards leaving the interface as is so that any tags on the network state can always be assumed to have data and error keys vs. having to confirm the existence of the keys before checking if they are null.

Is there a way to concisely check the existence and non-nullness of a key in a JS object?